File: sdk/nodejs/src/integrations/elevenlabs.ts
/**
 * ElevenLabs Conversational AI Integration for Hystersis - Node.js SDK
 *
 * Provides memory context for ElevenLabs conversational AI agents.
 *
 * @example
 * ```typescript
 * import { HystersisElevenLabsMemory } from 'hystersis/integrations/elevenlabs';
 *
 * const memory = new HystersisElevenLabsMemory({
 *   apiKey: 'your-api-key',
 *   baseUrl: 'https://api.hystersis.ai'
 * });
 *
 * // Get system prompt context
 * const context = await memory.getSystemPromptContext('user-123');
 *
 * // Store a conversation
 * await memory.storeConversation([
 *   { role: 'user', content: 'Tell me about my last order' },
 *   { role: 'assistant', content: 'Your last order was...' }
 * ], 'user-123');
 * ```
 */

import { HystersisClient as Hystersis } from '../index.js';

export interface ElevenLabsMemoryConfig {
  apiKey: string;
  baseUrl: string;
  contextLimit?: number;
}

/**
 * Memory context provider for ElevenLabs conversational AI.
 */
export class HystersisElevenLabsMemory {
  private client: Hystersis;
  private contextLimit: number;

  constructor(config: ElevenLabsMemoryConfig) {
    this.client = new Hystersis({
      baseUrl: config.baseUrl,
      apiKey: config.apiKey,
    });
    this.contextLimit = config.contextLimit ?? 5;
  }

  /**
   * Get memory context to prepend to the ElevenLabs system prompt.
   */
  async getSystemPromptContext(userId: string, query?: string): Promise<string> {
    try {
      let items: Array<Record<string, unknown>>;

      if (query) {
        const results = await this.client.memories.search(query, {
          user_id: userId,
          limit: this.contextLimit,
        });
        items = Array.isArray(results) ? results : [];
      } else {
        const results = await this.client.memories.list({
          user_id: userId,
          limit: this.contextLimit,
        });
        items = Array.isArray(results) ? results : [];
      }

      const texts = items
        .map((item) => (item.text as string) || (item.content as string) || '')
        .filter(Boolean)
        .map((t) => `- ${t}`);

      if (texts.length === 0) return '';

      return (
        'You have the following context about this user from previous conversations:\n' +
        texts.join('\n')
      );
    } catch {
      return '';
    }
  }

  /**
   * Store conversation messages in Hystersis memory.
   */
  async storeConversation(
    messages: Array<{ role: string; content: string }>,
    userId: string
  ): Promise<void> {
    for (const msg of messages) {
      if (!msg.content) continue;
      try {
        await this.client.memories.create({
          content: msg.content,
          user_id: userId,
          type: 'conversation',
          metadata: {
            source: 'elevenlabs',
            role: msg.role,
          },
        });
      } catch (error) {
        console.warn('Failed to store message:', error);
      }
    }
  }
}

File: sdk/python/hystersis/integrations/elevenlabs.py
"""
ElevenLabs Conversational AI Integration for Hystersis

This module provides memory context for ElevenLabs conversational AI agents,
enabling persistent memory across voice conversations.

Usage:
    from hystersis.integrations.elevenlabs import HystersisElevenLabsMemory

    memory = HystersisElevenLabsMemory(
        api_key="your-api-key",
        base_url="https://api.hystersis.ai"
    )

    # Get system prompt context for a user
    context = memory.get_system_prompt_context(user_id="user-123")

    # Store a conversation
    memory.store_conversation(
        messages=[
            {"role": "user", "content": "Tell me about my last order"},
            {"role": "assistant", "content": "Your last order was..."}
        ],
        user_id="user-123"
    )
"""

from typing import Any, Dict, List, Optional

import requests


class HystersisElevenLabsMemory:
    """
    Memory context provider for ElevenLabs conversational AI.

    Retrieves relevant memories to include in system prompts and stores
    conversation history for future context retrieval.

    Attributes:
        base_url: Base URL of the Hystersis API
        api_key: API key for authentication
        context_limit: Max number of memories to include in system prompt
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.hystersis.ai",
        context_limit: int = 5,
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.context_limit = context_limit

        self._session = requests.Session()
        self._session.headers.update({"X-API-Key": self.api_key})

    def _request(self, method: str, endpoint: str, **kwargs) -> requests.Response:
        """Make an HTTP request to the Hystersis API."""
        url = f"{self.base_url}{endpoint}"
        resp = self._session.request(method, url, **kwargs)
        resp.raise_for_status()
        return resp

    def get_system_prompt_context(
        self, user_id: str, query: Optional[str] = None
    ) -> str:
        """
        Get memory context to prepend to the ElevenLabs system prompt.

        Retrieves the most relevant memories for the user and formats them
        as a context block suitable for inclusion in a system prompt.

        Args:
            user_id: User identifier to retrieve memories for
            query: Optional search query; if None, retrieves recent memories

        Returns:
            Formatted context string for the system prompt
        """
        try:
            if query:
                resp = self._request(
                    "GET",
                    "/search",
                    params={
                        "q": query,
                        "user_id": user_id,
                        "limit": self.context_limit,
                    },
                )
            else:
                resp = self._request(
                    "GET",
                    "/memories",
                    params={
                        "user_id": user_id,
                        "limit": self.context_limit,
                    },
                )

            results = resp.json()
            items = results if isinstance(results, list) else results.get("results", results.get("memories", []))
        except Exception:
            return ""

        if not items:
            return ""

        memory_texts = []
        for item in items:
            text = item.get("text", "") or item.get("content", "")
            if text:
                memory_texts.append(f"- {text}")

        if not memory_texts:
            return ""

        return (
            "You have the following context about this user from previous conversations:\n"
            + "\n".join(memory_texts)
        )

    def store_conversation(
        self, messages: List[Dict[str, str]], user_id: str
    ) -> None:
        """
        Store conversation messages in Hystersis memory.

        Args:
            messages: List of message dicts with 'role' and 'content'
            user_id: User identifier
        """
        for msg in messages:
            content = msg.get("content", "")
            role = msg.get("role", "user")
            if not content:
                continue

            try:
                self._request(
                    "POST",
                    "/memories",
                    json={
                        "content": content,
                        "user_id": user_id,
                        "type": "conversation",
                        "metadata": {
                            "source": "elevenlabs",
                            "role": role,
                        },
                    },
                )
            except Exception as e:
                print(f"Warning: Failed to store message: {e}")

    def __repr__(self) -> str:
        return "HystersisElevenLabsMemory()"


__all__ = ["HystersisElevenLabsMemory"]
